### PR TITLE
fix(major): [sc-2502] Log a warning if opposite side of the connection does not respond to ping messages.

### PR DESCRIPTION
**Pull request by ser-0xff**
… does not respond to ping messages.

## Description

Log a warning if opposite side of the connection does not respond to ping messages.

## How Has This Been Tested?

Unit and manual tests.

## Minimal checklist:

- [X] I have performed a self-review of my own code 
- [ ] I have added `DocC` code-level documentation for any public interfaces exported by the package
- [ ] I have added unit and/or integration tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by blindspotbounty — APPROVED** [review]

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 1]
```diff
@@ -576,12 +589,52 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
...
+                    case let .clientForRemoteService(serviceChannel, _, _):
+                        channel = serviceChannel
+                    default:
+                        logger.error("Internal error: unexpected actor state \(actorInfo) for \(id)")
+                    }
+
+                    if let channel {
+                        let ptr = Self.ptr(for: channel)
+                        let channelInfo = self.channels[ptr]
+                        if var channelInfo {
```
Comment (dimlio) [inline Sources/DistributedSystem/DistributedSystem.swift:618]:
```suggestion
                        if var channelInfo = self.channels[ptr] {
```

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 2]
```diff
@@ -706,6 +764,7 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
                 }
                 return actor
             } else {
+                /*
```
Comment (dimlio) [inline Sources/DistributedSystem/DistributedSystem.swift:767]: Remove?

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 3]
```diff
@@ -806,16 +866,18 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
...
-            switch actorInfo {
-            case let .remoteClient(channel):
-                return channel
-            case let .remoteService(channel):
-                return channel
-            default:
-                fatalError("Internal error: invalid actor state \(actor.id): \(actorInfo)")
-            }
-        }()
+        var channel: Channel?
```
Comment (dimlio) [inline Sources/DistributedSystem/DistributedSystem.swift:869]:
```suggestion
        let channel: Channel? = switch actorInfo {
            case let .remoteClient(clientChannel): clientChannel
            case let .remoteService(serviceChannel): serviceChannel
            ...
        }
```

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 4]
```diff
@@ -181,9 +187,16 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
         case serviceForRemoteClient(Channel, AsyncStream<InvocationEnvelope>.Continuation, ManagedAtomic<UInt64>)
     }
 
+    struct ChannelInfo {
+        var bytesReceived = 0
```
Comment (dimlio) [inline Sources/DistributedSystem/DistributedSystem.swift:191]: Maybe some description of these fields?

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 5]
```diff
@@ -875,12 +937,24 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
         } catch {
             logger.error("\(channel.remoteAddressDescription): \(error)")
         }
+
+        let ptr = Self.ptr(for: channel)
+        lock.withLockVoid {
+            let channelInfo = self.channels[ptr]
```
Comment (dimlio) [inline Sources/DistributedSystem/DistributedSystem.swift:943]:
```swift
self.channels[ptr, default: ChannelInfo()].bytesReceived += bytesReceived
```
?

File: Sources/DistributedSystem/PingService.swift [hunk 6]
```diff
@@ -1,62 +1,68 @@
...
+                try await serviceEndpoint.pong()
+            } catch {
+                actorSystem.logger.error("\(error)")
+            }
+        }
     }
 
     distributed func pong() async throws {
-        await client.pong()
+        // actorSystem.logger.debug("\(id): pong")
```
Comment (dimlio) [inline Sources/DistributedSystem/PingService.swift:66]: Maybe move to `trace` instead of commenting?

**Review by dimlio — APPROVED** [review]

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 7]
```diff
@@ -149,13 +155,13 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
         case datacenter
     }
 
-    public static var logger = Logger(label: "ds")
-    var logger: Logger { Self.logger }
+    public var loggerBox = Box<Logger>(Logger(label: "ds"))
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift:158]: Will do it as soon the `@Box` property wrapper will be available :)

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 8]
```diff
@@ -181,9 +187,16 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
...
+        var bytesReceived = 0
+        var bytesReceivedCheckpoint = 0
+        var bytesReceivedTimeouts = 0
+        var connectionLossHandlers = [ConnectionLossHandler]()
+    }
+
     private var lock = Lock()
     private var actors: [EndpointIdentifier: ActorInfo] = [:]
-    private var connectionLossHandlers: [UnsafeMutableRawPointer: [ConnectionLossHandler]] = [:]
+    private var channels: [UnsafeRawPointer: ChannelInfo] = [:]
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift:199]: Would not like to add extra entity here... what is a purpose?

File: Sources/DistributedSystem/DistributedSystem.swift (hunk 4 above)
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift:191]: Well... is it really not clear from the name?

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 9]
```diff
@@ -576,12 +589,52 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
...
         Task {
             do {
-                try await serviceEndpoint.ping()
+                try await endpoint.ping()
+
+                self.lock.withLockVoid {
+                    let actorInfo = self.actors[id]
+                    guard let actorInfo else { return }
+
+                    var channel: Channel?
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift:598]: It does not work if we want to perform some call but not only return value from the 'case' clause.

File: Sources/DistributedSystem/DistributedSystem.swift (hunk 1 above)
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift:615]: Fixed.

File: Sources/DistributedSystem/DistributedSystem.swift (hunk 2 above)
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift:764]: Would like to leave it for now.

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 10]
```diff
@@ -369,24 +382,24 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
...
-            }()
+            let serviceEndpointID = self.duplicatedEndpointIdentifier.flatMap { (serviceEndpointType != PingServiceEndpoint.self) ? $0 : nil } ?? EndpointIdentifier(serviceID)
 
             if let channel {
                 self.lock.withLockVoid {
                     self.actors[serviceEndpointID] = .remoteService(channel)
                 }
+                self.sendCreateService(serviceName, serviceEndpointID, to: channel)
+            } else {
+                if (serviceEndpointType == PingServiceEndpoint.self) {
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift]: Fixed.

File: Sources/DistributedSystem/DistributedSystem.swift [hunk 11]
```diff
@@ -576,12 +589,52 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
...
 
-    private static func sendPing(to serviceEndpoint: PingServiceEndpoint, with eventLoop: EventLoop) {
+    private func sendPing(to endpoint: PingEndpoint, id: EndpointIdentifier, with eventLoop: EventLoop) {
         Task {
             do {
-                try await serviceEndpoint.ping()
+                try await endpoint.ping()
+
+                self.lock.withLockVoid {
+                    let actorInfo = self.actors[id]
```
[the remark this answers is not in the document]
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift]: Fixed.

File: Sources/DistributedSystem/DistributedSystem.swift (hunk 5 above)
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift]: Fixed.

File: Sources/DistributedSystem/DistributedSystem.swift (hunk 3 above)
Comment (ser-0xff) [reply Sources/DistributedSystem/DistributedSystem.swift]: Fixed.

File: Sources/DistributedSystem/PingService.swift (hunk 6 above)
Comment (ser-0xff) [reply Sources/DistributedSystem/PingService.swift]: Fixed.

**Pull request merged**